include/kamping/distributed_graph_communicator.hpp

```diff
@@ -0,0 +1,353 @@
+
+        MPI_Comm mpi_graph_comm;
+
+        MPI_Dist_graph_create_adjacent(
+            comm,
+            in_degree_signed(),
+            in_ranks().data(),
+            in_weights,
+            out_degree_signed(),
+            out_ranks().data(),
+            out_weights,
+            MPI_INFO_NULL,
+            false,
+            &mpi_graph_comm
+        );
+        return mpi_graph_comm;
+    }
+
+private:
+    SpanType                _in_ranks;    ///< View on in-going edges.
+    SpanType                _out_ranks;   ///< View on out-going edges.
+    std::optional<SpanType> _in_weights;  ///< View on in-going edge weights.
+    std::optional<SpanType> _out_weights; ///< View on out-going edge weights.
+};
+
+namespace internal {
+/// @brief Returns whether a given range of edges is weighted or not at compile time
+/// @tparam EdgeRange Range type to be checked.
+template <typename EdgeRange>
+constexpr bool are_edges_weighted() {
+    using EdgeType = typename EdgeRange::value_type;
+    return !std::is_integral_v<EdgeType>;
+}
+} // namespace internal
+
+/// @brief A Distributed communication graph. Each vertex of the graph corresponds to a rank and each edge (i,j) connect
+/// two ranks i and j which can communicate with each other. Note that MPI allow multiple edges between the same ranks i
+/// and j, i.e. the distributed communication graph can be a multi-graph. Each rank holds its local view on the
+/// communication graph, i.e., it knows its neighboring vertices/ranks.
+template <template <typename...> typename DefaultContainer = std::vector>
+class CommunicationGraph {
+public:
+    /// @brief Default constructor.
+    CommunicationGraph() = default;
+
+    /// @brief Constructs a communication graph based on a range of in-going and out-going edges which might be
+    /// weighted. A unweighted edge is simply an integer, wherea a weighted edge is a pair-like object consisting of two
+    /// integers which can be decomposed by a structured binding via `auto [rank, weight] = weighted_edge;`.
+    ///
+    /// @tparam InEdgeRange Range type of in-going edges.
+    /// @tparam InEdgeRange Range type of out-going edges.
+    /// @param in_edges Range of in-going edges.
+    /// @param out_edges Range of out-going edges.
+    ///
+    template <typename InEdgeRange, typename OutEdgeRange>
+    CommunicationGraph(InEdgeRange const& in_edges, OutEdgeRange const& out_edges) {
+        constexpr bool are_in_edges_weighted  = internal::are_edges_weighted<InEdgeRange>();
+        constexpr bool are_out_edges_weighted = internal::are_edges_weighted<OutEdgeRange>();
+        static_assert(
+            are_in_edges_weighted == are_out_edges_weighted,
+            "Weight status of in and out edges is different!"
+        );
+        if constexpr (are_in_edges_weighted) {
+            auto get_rank = [](auto const& edge) {
+                auto const& [rank, _] = edge;
+                return static_cast<int>(rank);
+            };
```

Comment:
move `if constexpr` inside of get rank and add additional if for weights. Makes the code easier to read.